discovery/azure/azure.go

```diff
@@ -539,7 +555,7 @@ func mapFromVMScaleSetVM(vm compute.VirtualMachineScaleSetVM, scaleSetName strin
 	}
 }
 
-func (client *azureClient) getNetworkInterfaceByID(ctx context.Context, networkInterfaceID string) (*network.Interface, error) {
+func (client *azureClient) getNetworkInterfaceByID(ctx context.Context, networkInterfaceID string) (*network.Interface, *autorest.DetailedError) {
```

Comment:
Functions/methods should always return errors as the abstract error interface.

```suggestion
func (client *azureClient) getNetworkInterfaceByID(ctx context.Context, networkInterfaceID string) (*network.Interface, error) {
```

Reply:
I agree with David here. We are hugely dependent on Azure's autorest package, and I prefer to stick to its types. I think the issue is that packaged forced to us, I don't think we should loop into `error`.

Reply:
Can we narrow down the error to ignore in getNetworkInterfaceByID? This particular error would ignorer 404 in 3 different calls.

Reply:
I can change it to just check and ignore a 404 on the last method, `client.nic.GetResponder()`. But I need a way to indicate to the caller this should be ignored. Are you OK with just returning `nil,nil` in that case? Seems a little odd to me. Otherwise, I'd need to return a different error and do a check, although slightly different, where it's called.

Reply:
In general, if a function returns a value that the caller can't actually use, you want to model that as an error. Another way of saying that is that a `nil` error means the other return values are valid. So as a strawman

```
func (c *azureClient) getNetworkInterfaceByID(ctx context.Context, id string) (*network.Interface, error) {
    ...
    switch {
    case success:
        return &network.Interface{...}, nil
    case code was 404:
        return nil, ErrNotFound
    case other failure:
        return nil, fmt.Errorf("other failure: %w", err)
    }
}
```

Callers can program against this error return via normal mechanisms.

```go
iface, err := c.getNetworkInterfaceByID(ctx, id)
switch {
case err == nil:
    // do whatever you do in a success case
case errors.Is(err, ErrNotFound):
    // do whatever you do when it's not found
case err != nil:
    // do whatever you do in a failure case
}
```

Reply:
If callers need to access specific details from the error, then you can use types and `errors.As`.

```go
    case code was 404:
        return nil, &DetailedError{...}
```

```go
var detailedErr *DetailedError
iface, err := c.getNetworkInterfaceByID(ctx, id)
switch {
case errors.As(err, &detailedErr):
    // use detailedErr
```

Reply:
Yes, I found similar patterns when researching custom errors after your first post. I asked, because I didn't see any advantage to creating custom errors over the current implementation of just returning a `DetailedError`.

Reply:
> It’s a good idea for functions that return errors always to use the error type in their signature (as we did above) rather than a concrete type such as *MyError, to help guarantee the error is created correctly. As an example, os.Open returns an error even though, if not nil, it’s always of concrete type *os.PathError.

[reference](https://go.dev/doc/faq#nil_error), [reference](https://staticcheck.io/docs/checks#SA4023), etc.

Reply:
Thank you for the references,they were very helpful.

Reply:
I've updated this method to return the `error` interface with different concrete types. The 404 check is performed here instead of by the caller.